### DISCUSSION TITLE
Converting netCDF to GRIB2

### DISCUSSION THREAD

**Discussion by johnlwagner** [dtcenter/METplus#1015] [File I/O] [answered]
Has anyone converted a netCDF file produced by MET back to a GRIB2 file?  I'm currently testing grids that have been created by series analysis.  I've tried using iris and cdo to convert back to GRIB2, but both seem to have issues with the lat/lon dimensions.
With cdo, there seem to be issues with the dimensions and variables being called "lat" and "lon" instead of "latitude" and "longitude".  I am able to use ncatted and ncrename to correct the "lat" settings, but never the "lon" variable.

dimensions:
        lon = 2345 ;
        lat = 1597 ;
variables:
        float **latitude**(lat, lon) ;
                latitude:standard_name = "latitude" ;
                latitude:long_name = "latitude" ;
                latitude:units = "degrees_north" ;
        float **lon**(lat, lon) ;
                lon:standard_name = "longitude" ;
                lon:long_name = "longitude" ;
                lon:units = "degrees_east" ;

The error cdo gives is
ECCODES ERROR   :  concept: no match for cfName=longitude
ECCODES ERROR   :  concept: input handle edition=2
!!! grib_set_string(    grib_handle* h, "cfName", "longitude") !!!
Warning (gribapiDefParam): grib_api: No match for cfName=longitude

Error (gribapiDefGrid): Unsupported grid type: generic

Even using cdo -chName will not change "lon" to "longitude".
Iris seems to have issues with the lat and lon variables being 2-dimensional.  With WCOSS dev access restricted at the moment, I cannot access the exact error.

If anyone knows of another utility to convert these grids to GRIB2, please let me know.  If you know of any other utilities that can be used to force the netCDF variables to change, that would also be appreciated.

The series analysis output grids can be found in my Google Drive folder here:
https://drive.google.com/drive/folders/1tr2blv_JwePMWNwyfqJhwA3_dIffM9OG?usp=sharing

**Comment by JohnHalleyGotway** [comment, marked as answer]
@johnlwagner, I wanted to let you know that no one within the DTC, to my knowledge, has converted MET NetCDF output into GRIB2 files... although we've never really been motivated to do so. What is the motivation for this work? Do you need GRIB2 output for display on AWIPS?

I see that you're trying to use the CDO for this conversion but are having trouble renaming the 'lon' variable to 'longitude'. I typically use the 'nco' utility named 'ncrename' for this sort thing.
```
ncrename -d lat,latitude -d lon,longitude -v lat,latitude -v lon,longitude series_analysis_UPPER_AIR_TMP_120000L_20120409_120000V.nc
```
However that produced an error:
```
NetCDF: Can't define dimensional metadata
```
One option, just for testing, is running ncdump followed by ncgen:
```
ncdump series_analysis_UPPER_AIR_TMP_120000L_20120409_120000V.nc | sed 's/lat/latitude/g' | sed 's/lon/longitude/g' > sa.dump
ncgen -o series_analysis_UPPER_AIR_TMP_120000L_20120409_120000V_MOD.nc sa.dump
```
And that'll do the trick. However, the real tricky part with writing GRIB2 files is getting the metadata and grid definition correct. If your data is on a lat/lon grid, that'd be relatively simple. But if it's projected on Lambert Conformal, Polar Stereographic, or Mercator that that'd be trickier.

If I were try to do this, I'd explore options in python. The pygrib, pynio, and grib2io packages sound somewhat promising.

**Reply by johnlwagner** [reply]
Thanks John.  The ncdump/ncgen step did work.  Now I get a new set of errors from cdo to work through.
I'll look into the python options.  That how I got started down the path of iris and cdo.
To answer your question, the USCG asked us for gridded NDFD wind speed errors.  I believe they work with grib2 files.  I'm not sure if they're using them in AWIPS or not.

**Reply by JohnHalleyGotway** [reply]
Ah OK, sounds interesting. I can see how writing a python utility in METplus to handle the conversion of Series-Analysis NetCDF output to a GRIB2 file would be useful. But, as you're finding, it'd take some time and effort to do that well. Please let us know if you do have some success, and we could consider if/where/how to incorporate your work back into METplus.

**Reply by JohnHalleyGotway** [reply]
@johnlwagner I'm marking this discussion as "answered" to move it out of the support "inbox". But please feel free to add more comments in the futre.

**Comment by johnlwagner** [comment]
Sounds good.  Thanks John.
We're still testing this when we have time.  If we do find a solution,
we'll let you know.

On Mon, Aug 9, 2021 at 1:31 PM johnhg ***@***.***> wrote:

> @johnlwagner <https://github.com/johnlwagner> I'm marking this discussion
> as "answered" to move it out of the support "inbox". But please feel free
> to add more comments in the futre.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/METplus/discussions/1015#discussioncomment-1150590>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGZ7TYMAXRGHCW4S422TK6LT4AGHDANCNFSM5AX2D2AQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>


-- 
John Wagner
Verification Task Lead
NOAA/National Weather Service
Meteorological Development Laboratory
Digital Forecast Services Division
SSMC2 Room 10106
Silver Spring, MD 20910
(301) 427-9471 (office)
(908) 902-4155 (cell/text)
